worlds/pokemon_crystal/rom.py

```diff
@@ -0,0 +1,425 @@
+import copy
+import os
+from typing import TYPE_CHECKING
+
+import bsdiff4
+
+from settings import get_settings
+from worlds.Files import APProcedurePatch, APTokenMixin, APTokenTypes, APPatchExtension
+from .data import data
+from .items import item_const_name_to_id
+from .utils import convert_to_ingame_text
+
+if TYPE_CHECKING:
+    from . import PokemonCrystalWorld
+
+
+class PokemonCrystalAPPatchExtension(APPatchExtension):
+    game = "Pokemon Crystal"
+
+    @staticmethod
+    def apply_bsdiff4(caller: APProcedurePatch, rom: bytes, patch: str):
+        revision_address = data.rom_addresses["AP_ROM_Revision"]
+        rom_bytes = bytearray(rom)
+        if rom_bytes[revision_address] == 1:
+            if "basepatch11.bsdiff4" not in caller.files:
+                raise Exception("This patch was generated without support for Pokemon Crystal V1.1 ROM. "
+                                "Please regenerate with a newer APWorld version or use a V1.0 ROM")
+            return bsdiff4.patch(rom, caller.get_file("basepatch11.bsdiff4"))
+        return bsdiff4.patch(rom, caller.get_file(patch))
+
+
+class PokemonCrystalProcedurePatch(APProcedurePatch, APTokenMixin):
+    game = "Pokemon Crystal"
+    hash = "9f2922b235a5eeb78d65594e82ef5dde"
+    patch_file_ending = ".apcrystal"
+    result_file_ending = ".gbc"
+
+    procedure = [
+        ("apply_bsdiff4", ["basepatch.bsdiff4"]),
+        ("apply_tokens", ["token_data.bin"])
+    ]
+
+    @classmethod
+    def get_source_data(cls) -> bytes:
+        return get_base_rom_as_bytes()
+
+
+def generate_output(world: "PokemonCrystalWorld", output_directory: str, patch: PokemonCrystalProcedurePatch) -> None:
+    item_texts = []
+    for location in world.multiworld.get_locations(world.player):
+        if location.address is None:
+            continue
+
+        if location.item and location.item.player == world.player:
+            item_id = location.item.code
+            item_id = item_id - 256 if item_id > 256 else item_id
+            write_bytes(patch, [item_id], location.rom_address)
+        else:
+            # for in game text
+            item_flag = location.address
+            player_name = world.multiworld.player_name[location.item.player].upper()
+            item_name = location.item.name.upper()
+            item_texts.append([player_name, item_name, item_flag])
+
+            write_bytes(patch, [item_const_name_to_id("AP_ITEM")], location.rom_address)
```

Comment:
`item_const_name_to_id("AP_ITEM")` could be saved to a local variable, since it iterates over all Crystal items every time it's called. Or have `item_const_name_to_id` use `@cache` as suggested above.